elements/castbar.lua

```diff
@@ -83,443 +85,290 @@ A default texture will be applied to the StatusBar and Texture widgets if they d
     Castbar.SafeZone = SafeZone
     self.Castbar = Castbar
 --]]
+
 local _, ns = ...
 local oUF = ns.oUF
 
-local GetNetStats = GetNetStats
-local GetTime = GetTime
-local UnitCastingInfo = UnitCastingInfo
-local UnitChannelInfo = UnitChannelInfo
-
-local function updateSafeZone(self)
-	local safeZone = self.SafeZone
-	local width = self:GetWidth()
-	local _, _, _, ms = GetNetStats()
-
-	local safeZoneRatio = (ms / 1e3) / self.max
-	if(safeZoneRatio > 1) then
-		safeZoneRatio = 1
-	end
-
-	safeZone:SetWidth(width * safeZoneRatio)
+local function resetAttributes(self)
+	self.castID = nil
+	self.casting = nil
+	self.channeling = nil
+	self.notInterruptible = nil
+	self.spellID = nil
 end
 
-local function UNIT_SPELLCAST_START(self, event, unit)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+local function CastStart(self, event, unit)
+	if(self.unit ~= unit) then return end
 
 	local element = self.Castbar
-	local name, text, texture, startTime, endTime, _, castID, notInterruptible, spellID = UnitCastingInfo(unit)
+
+	local name, _, texture, startTime, endTime, isTradeSkill, castID, notInterruptible, spellID = UnitCastingInfo(unit)
+	event = 'UNIT_SPELLCAST_START'
 	if(not name) then
-		return element:Hide()
+		name, _, texture, startTime, endTime, isTradeSkill, notInterruptible, spellID = UnitChannelInfo(unit)
+		event = 'UNIT_SPELLCAST_CHANNEL_START'
 	end
 
-	endTime = endTime / 1e3
-	startTime = startTime / 1e3
-	local max = endTime - startTime
+	if(not name or (isTradeSkill and element.hideTradeSkills)) then
+		resetAttributes(element)
+		element:Hide()
 
-	element.castID = castID
-	element.duration = GetTime() - startTime
-	element.max = max
+		return
+	end
+
+	endTime = endTime / 1000
+	startTime = startTime / 1000
+
+	element.max = endTime - startTime
+	element.startTime = startTime
 	element.delay = 0
-	element.casting = true
+	element.casting = event == 'UNIT_SPELLCAST_START'
+	element.channeling = event == 'UNIT_SPELLCAST_CHANNEL_START'
 	element.notInterruptible = notInterruptible
 	element.holdTime = 0
+	element.castID = castID
 	element.spellID = spellID
 
-	element:SetMinMaxValues(0, max)
+	if(element.casting) then
+		element.duration = GetTime() - startTime
+	else
+		element.duration = endTime - GetTime()
+	end
+
+	element:SetMinMaxValues(0, element.max)
 	element:SetValue(0)
 
-	if(element.Text) then element.Text:SetText(text) end
-	if(element.Icon) then element.Icon:SetTexture(texture) end
+	if(element.Icon) then element.Icon:SetTexture(texture or [[Interface\ICONS\INV_Misc_QuestionMark]]) end
```

Comment:
Use a constant for the fallback texture

Reply:
Thought so, even did so initially, but... >_>